File: ios/ChessClock/Game/PawnIcon.swift
import SwiftUI

/// Silhouette du pion — tracé calqué sur les maquettes officielles (tête ronde, collerette,
/// taille cintrée, base évasée à bord plat), boîte de référence 100 x 140.
private struct PawnShape: Shape {
    func path(in rect: CGRect) -> Path {
        let u = rect.height / 140
        var path = Path()
        path.move(to: CGPoint(x: 50 * u, y: 0))
        path.addCurve(to: CGPoint(x: 74 * u, y: 20 * u), control1: CGPoint(x: 66 * u, y: 0), control2: CGPoint(x: 74 * u, y: 10 * u))
        path.addCurve(to: CGPoint(x: 70 * u, y: 36 * u), control1: CGPoint(x: 74 * u, y: 28 * u), control2: CGPoint(x: 70 * u, y: 34 * u))
        path.addCurve(to: CGPoint(x: 83 * u, y: 54 * u), control1: CGPoint(x: 78 * u, y: 40 * u), control2: CGPoint(x: 83 * u, y: 44 * u))
        path.addLine(to: CGPoint(x: 83 * u, y: 73 * u))
        path.addCurve(to: CGPoint(x: 63 * u, y: 77 * u), control1: CGPoint(x: 83 * u, y: 76 * u), control2: CGPoint(x: 76 * u, y: 76 * u))
        path.addCurve(to: CGPoint(x: 80 * u, y: 116 * u), control1: CGPoint(x: 68 * u, y: 85 * u), control2: CGPoint(x: 74 * u, y: 95 * u))
        path.addCurve(to: CGPoint(x: 91 * u, y: 131 * u), control1: CGPoint(x: 85 * u, y: 122 * u), control2: CGPoint(x: 91 * u, y: 126 * u))
        path.addLine(to: CGPoint(x: 91 * u, y: 140 * u))
        path.addLine(to: CGPoint(x: 9 * u, y: 140 * u))
        path.addLine(to: CGPoint(x: 9 * u, y: 131 * u))
        path.addCurve(to: CGPoint(x: 20 * u, y: 116 * u), control1: CGPoint(x: 9 * u, y: 126 * u), control2: CGPoint(x: 15 * u, y: 122 * u))
        path.addCurve(to: CGPoint(x: 37 * u, y: 77 * u), control1: CGPoint(x: 26 * u, y: 95 * u), control2: CGPoint(x: 32 * u, y: 85 * u))
        path.addCurve(to: CGPoint(x: 17 * u, y: 73 * u), control1: CGPoint(x: 24 * u, y: 76 * u), control2: CGPoint(x: 17 * u, y: 76 * u))
        path.addLine(to: CGPoint(x: 17 * u, y: 54 * u))
        path.addCurve(to: CGPoint(x: 30 * u, y: 36 * u), control1: CGPoint(x: 17 * u, y: 44 * u), control2: CGPoint(x: 22 * u, y: 40 * u))
        path.addCurve(to: CGPoint(x: 26 * u, y: 20 * u), control1: CGPoint(x: 26 * u, y: 34 * u), control2: CGPoint(x: 26 * u, y: 28 * u))
        path.addCurve(to: CGPoint(x: 50 * u, y: 0), control1: CGPoint(x: 26 * u, y: 10 * u), control2: CGPoint(x: 34 * u, y: 0))
        path.closeSubpath()
        return path
    }
}

/// Pion d'échecs vectoriel — exception assumée au principe "aucune pièce dessinée" de la
/// planche, à la demande explicite du produit. [contourColor] : toujours le ton opposé au
/// fond du pion — sur son propre camp, le fond se fond avec le fond de sa zone, seul ce
/// contour le rend visible.
struct PawnIcon: View {
    let fillColor: Color
    let contourColor: Color
    let height: CGFloat

    private var width: CGFloat { height * (100 / 140) }

    var body: some View {
        ZStack {
            PawnShape().fill(fillColor)
            PawnShape().stroke(contourColor, lineWidth: height * (2 / 140))
        }
        .frame(width: width, height: height)
    }
}

File: android/app/src/main/kotlin/dev/alek6dev/chessclock/ui/game/PawnIcon.kt
package dev.alek6dev.chessclock.ui.game

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.size
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.unit.Dp

/**
 * Pion d'échecs vectoriel — exception assumée au principe "aucune pièce dessinée" de la
 * planche, à la demande explicite du produit pour identifier visuellement les chronos par
 * camp. Tracé calqué sur les maquettes officielles (tête ronde, collerette, taille cintrée,
 * base évasée à bord plat), dans une boîte de référence 100 x 140 mise à l'échelle par
 * [height].
 *
 * [contourColor] : toujours le ton opposé au fond du pion (ivoire pour un pion encre, encre
 * pour un pion ivoire). Sur son propre camp, le fond du pion se fond avec le fond de sa zone
 * (même couleur) — seul ce contour le rend visible.
 */
@Composable
fun PawnIcon(fillColor: Color, contourColor: Color, height: Dp, modifier: Modifier = Modifier) {
    val width = height * (100f / 140f)
    Canvas(modifier = modifier.size(width = width, height = height)) {
        val u = size.height / 140f
        val body = Path().apply {
            moveTo(50 * u, 0f)
            cubicTo(66 * u, 0f, 74 * u, 10 * u, 74 * u, 20 * u)
            cubicTo(74 * u, 28 * u, 70 * u, 34 * u, 70 * u, 36 * u)
            cubicTo(78 * u, 40 * u, 83 * u, 44 * u, 83 * u, 54 * u)
            lineTo(83 * u, 73 * u)
            cubicTo(83 * u, 76 * u, 76 * u, 76 * u, 63 * u, 77 * u)
            cubicTo(68 * u, 85 * u, 74 * u, 95 * u, 80 * u, 116 * u)
            cubicTo(85 * u, 122 * u, 91 * u, 126 * u, 91 * u, 131 * u)
            lineTo(91 * u, 140 * u)
            lineTo(9 * u, 140 * u)
            lineTo(9 * u, 131 * u)
            cubicTo(9 * u, 126 * u, 15 * u, 122 * u, 20 * u, 116 * u)
            cubicTo(26 * u, 95 * u, 32 * u, 85 * u, 37 * u, 77 * u)
            cubicTo(24 * u, 76 * u, 17 * u, 76 * u, 17 * u, 73 * u)
            lineTo(17 * u, 54 * u)
            cubicTo(17 * u, 44 * u, 22 * u, 40 * u, 30 * u, 36 * u)
            cubicTo(26 * u, 34 * u, 26 * u, 28 * u, 26 * u, 20 * u)
            cubicTo(26 * u, 10 * u, 34 * u, 0f, 50 * u, 0f)
            close()
        }
        drawPath(body, color = fillColor)
        drawPath(body, color = contourColor, style = Stroke(width = 2 * u))
    }
}
